src/algebra/algebra/restrict_scalars.lean

```diff
@@ -0,0 +1,145 @@
+typeclass instead.
+
+## Main definitions
+
+* `restrict_scalars R S M`: the `S`-module `M` viewed as an `R` module when `S` is an `R`-algebra.
+* `restrict_scalars.linear_equiv : restrict_scalars R S M ≃ₗ[S] M`: the equivalence as an
+  `S`-module between the restricted and origianl space.
+* `restrict_scalars.alg_equiv : restrict_scalars R S A ≃ₐ[S] A`: the equivalence as an `S`-algebra
+   between the restricted and original space.
+
+## See also
+
+There are many similarly-named definitions elsewhere which do not refer to this type alias. These
+refer to restricting the scalar type in a bundled type, such as from `A →ₗ[R] B` to `A →ₗ[S] B`:
+
+* `linear_map.restrict_scalars`
+* `linear_equiv.restrict_scalars`
+* `alg_hom.restrict_scalars`
+* `alg_equiv.restrict_scalars`
+* `submodule.restrict_scalars`
+* `subalgebra.restrict_scalars`
+-/
+
+variables (R S M A : Type*)
+
+/-- If we put an `R`-algebra structure on a semiring `S`, we get a natural equivalence from the
+category of `S`-modules to the category of representations of the algebra `S` (over `R`). The type
+synonym `restrict_scalars` is essentially this equivalence.
+
+Warning: use this type synonym judiciously! Consider an example where we want to construct an
+`R`-linear map from `M` to `S`, given:
+```lean
+variables (R S M : Type*)
+variables [comm_semiring R] [semiring S] [algebra R S] [add_comm_monoid M] [module S M]
+```
+With the assumptions above we can't directly state our map as we have no `module R M` structure, but
+`restrict_scalars` permits it to be written as:
+```lean
+-- an `R`-module structure on `M` is provided by `restrict_scalars` which is compatible
+example : restrict_scalars R S M →ₗ[R] S := sorry
+```
+However, it is usually better just to add this extra structure as an argument:
+```lean
+-- an `R`-module structure on `M` and proof of its compatibility is provided by the user
+example [module R M] [is_scalar_tower R S M] : M →ₗ[R] S := sorry
+```
+The advantage of the second approach is that it defers the duty of providing the missing typeclasses
+`[module R M] [is_scalar_tower R S M]`. If some concrete `M` naturally carries these (as is often
+the case) then we have avoided `restrict_scalars` entirely. If not, we can pass
+`restrict_scalars R S M` later on instead of `M`.
+
+Note that this means we almost always want to state definitions and lemmas in the language of
+`is_scalar_tower` rather than `restrict_scalars`.
+
+An example of when one might want to use `restrict_scalars` would be if one has a vector space
+over a field of characteristic zero and wishes to make use of the `ℚ`-algebra structure. -/
+@[nolint unused_arguments]
+def restrict_scalars (R S M : Type*) : Type* := M
+
+instance [I : inhabited M] : inhabited (restrict_scalars R S M) := I
+
+instance [I : add_comm_monoid M] : add_comm_monoid (restrict_scalars R S M) := I
+
+instance [I : add_comm_group M] : add_comm_group (restrict_scalars R S M) := I
```

Comment:
Can these instances be `@[derive]`d?

Reply:
No, because they're conditional on properties of `M`